.github/workflows/check_task.yml

```
@@ -12,8 +10,82 @@ on:
-  build:
+  check-affected-files:
+    name: Check affected files
+    runs-on: ubuntu-latest
+    outputs:
+      affected_file: ${{ steps.check-affected-files.outputs.affected_file }}
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v45
+
+      - name: Check affected files
+        id: check-affected-files
+        env:
+          # get all affected files (created, modified, deleted, etc.)
+          FILES: ${{ steps.changed-files.outputs.all_changed_and_modified_files }}
+          FILES_COUNT: ${{ steps.changed-files.outputs.all_changed_and_modified_files_count }}
+        run: |
+          for file in "$FILES"; do
+            echo "Info: $file was changed"
+          done
+          if [ "$FILES_COUNT" -ne 1 ]; then
+            echo "Error: More than one file affected ($FILES_COUNT)"
+            exit 1
+          fi
+          if [[ "$FILES" =~ ^contributions\/(executable-tutorial|feedback|open-source)\/[^\/]+\/README\.md$ ]]; then
+            echo "Info: Matches async task"
+          elif [[ "$FILES" =~ ^contributions\/(demo|presentation|scientific-paper)\/week[2-7]\/[^\/]+\/README\.md$ ]]; then
+            echo "Info: Matches sync task"
+          else
+            echo "Error: File in wrong directory ($FILES)"
+            exit 1
+          fi
+
+          echo "affected_file=$FILES" >> "$GITHUB_OUTPUT"
+
+  check-pr-description:
+    name: Check whether Pull Request description matches affected file
+    needs: check-affected-files
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: Install wdiff
+        # we use wdiff (word-diff) instead of plain diff (line-based) to ensure
+        # that we only enforce *content* to be the same rather than specific
+        # file formatting that would otherwise be ignored when the markdown is
+        # rendered. e.g., wdiff allows the README to break lines at col 80 for
+        # readability while still having the PR description without mid-sentence
+        # breaks (since GitHub descriptions don't support them)
+        run: |
+          sudo apt update
+          sudo apt install -y wdiff
+
+      - name: Check PR description
+        env:
+          AFFECTED_FILE: ${{ needs.check-affected-files.outputs.affected_file }}
+          PR_BODY: ${{ github.event.pull_request.body }}
+        run: |
+          BODY_FILE="$(mktemp)"
+          echo "$PR_BODY" > $BODY_FILE
+          if wdiff "$AFFECTED_FILE" "$BODY_FILE"; then
+            echo "Info: PR description matches affected file"
+          else
+            echo "Error: PR description does not match affected file!"
+            exit 1
+          fi
+
+  check-canvas:
+    name: Check proposal is compatible with previous student task registrations
```

Unsure whether we should require our new check to pass before running the existing Python script.
```suggestion
  check-canvas:
    name: Check proposal is compatible with previous student task registrations
    needs: check-pr-description
```
For now, we have opted not to specify a dependency so that the jobs can run in parallel and be faster for "good PRs", which should be the majority in volume, even if "bad PRs" may suffer slightly since the last job is not aborted even if another one fails first. In any case, the scale is arguably negligible; all the checks are relatively quick.

---

I agree here. We can run them independently.